### ISSUE TITLE
StaticResourceResolver resolves index.html only from the root

### ISSUE THREAD

**Issue by klu2** [micronaut-projects/micronaut-core#409]
As described in https://docs.micronaut.io/latest/guide/index.html#staticResources, static resource resolving also resolves index.html files automatically. 

>  "In the above example a request to /static will attempt to retrieve src/main/resources/public/index.html."

What is not obvious there, is that this only works for the root folder of that static context. Taking the the example from above, if I add an `src/main/resources/public/sub/index.html`, then requesting `/static/sub or /static/sub/` returns a 404, whereas `/static/sub/index.html` is working fine.

Problem is the `StaticResourceResolver`, who appends the `index.html` only if the relative path of the resource loader is empty:

```
 public Optional<URL> resolve(String path) {
        if (!loaders.isEmpty() && pathMatcher.matches(mapping, path)) {
            path = pathMatcher.extractPathWithinPattern(mapping, path);
            //A request to the root of the mapping
            if (StringUtils.isEmpty(path)) {
                path = "index.html";
            }
            if (path.startsWith("/")) {
                path = path.substring(1);
            }
            for (ResourceLoader loader : loaders) {
                Optional<URL> resource = loader.getResource(path);
                if (resource.isPresent()) {
                    return resource;
                }
            }
        }
        return Optional.empty();
```

If would be quite easy to submit a pull request also to append the index.html to any subdirectories (by finding out if the last part of the path is a file or a directory) but I first wanted to discuss that with you if there is any intention behind the current implementation

**Comment by klu2** [comment]
what if a trailing slash ('/') could be the designator to append an index.html? that's not 100% clean, I know, as usually a webserver would also accept www.mycomain.com/sub and serve /sub/index.html then, but I think accepting the trailing slash is at least better than what we have now. what do you think?

**Comment by jameskleeh** [comment]
@klu2 I don't think it's a good idea to change the behavior of a response based on whether a slash is appended to the URL. It would lead to incorrect behavior if `sub` was a file and the request was issued to `/sub/`. The core issue is that the Jar resources don't provide any sort of API to determine if the resource is a directory

**Comment by klu2** [comment]
Agree. Then we should at least mention that more explicitely in the documentation, that detection of index.html only works on the root folder of the static resources, until we have a better solution here.
